parlai/chat_service/core/socket.py

```diff
@@ -0,0 +1,173 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+import errno
+import json
+import logging
+import threading
+import time
+import websocket
+import parlai.chat_service.utils.logging as log_utils
+
+SOCKET_TIMEOUT = 6
+
+
+# Socket handler
+class ChatServiceMessageSocket:
+    """
+    ChatServiceMessageSocket is a wrapper around websocket to forward messages from the
+    remote server to the ChatServiceManager.
+    """
+
+    def __init__(self, server_url, port, message_callback):
+        """
+        server_url:           url at which the server is to be run
+        port:                 port for the socket to operate on
+        message_callback:     function to be called on incoming message objects
+                              format: message_callback(self, data)
+        """
+        self.server_url = server_url
+        self.port = port
+        self.message_callback = message_callback
+
+        self.ws = None
+        self.last_pong = None
+        self.alive = False
+
+        # initialize the state
+        self.listen_thread = None
+
+        # setup the socket
+        self.keep_running = True
+        self._setup_socket()
+
+    def _safe_send(self, data, force=False):
+        if not self.alive and not force:
+            # Try to wait a second to send a packet
+            timeout = 1
+            while timeout > 0 and not self.alive:
+                time.sleep(0.1)
+                timeout -= 0.1
```

Comment:
interesting

Reply:
hmm i didnt even write this lol

Reply:
Yeaaah, I wrote it before learning about threading events. A better implementation would be to have `self.alive` be a `threading.Event` and call `self.alive.wait(1)` here, but there's no _need_ to update this as it would also require all `self.alive` checks to expect an `Event` rather than a bool.